packages/terra-clinical-onset-picker/src/OnsetPicker.jsx

```diff
@@ -0,0 +1,428 @@
+
+  /**
+   * Name of the granularity select. The name should be unique.
+   */
+  granularitySelectName: PropTypes.string.isRequired,
+
+  /**
+   * A callback function to execute when a granularity is selected.
+   * The first parameter is the event. The second parameter is the changed granularity value.
+   */
+  granularitySelectOnChange: PropTypes.func,
+
+  /**
+   * The precision of the onset date. ABOUT, BEFORE, AFTER, and UNKNOWN are accepted.
+   */
+  precision: PropTypes.oneOf(['ABOUT', 'BEFORE', 'AFTER', 'UNKNOWN']),
+
+  /**
+   * Name of the precision select. The name should be unique.
+   */
+  precisionSelectName: PropTypes.string.isRequired,
+
+  /**
+   * A callback function to execute when a precision is selected.
+   * The first parameter is the event. The second parameter is the changed precision value.
+   */
+  precisionSelectOnChange: PropTypes.func,
+
+  /**
+   * The ISO 8601 string representation of the onset date to view/modify. Defaults to current date.
+   */
+  onsetDate: PropTypes.string,
+
+  /**
+   * Name of the onsetDate input. The name should be unique.
+   */
+  onsetDateInputName: PropTypes.string.isRequired,
+
+  /**
+   * A callback function to execute when a onsetDate is changed.
+   * The first parameter is the event. The second parameter is the changed onsetDate value.
+   */
+  onsetDateInputOnChange: PropTypes.func,
+};
+
+const defaultProps = {
+  birthdate: undefined,
+  granularity: 'MONTH',
+  granularitySelectName: undefined,
+  granularitySelectOnChange: undefined,
+  precision: 'ABOUT',
+  precisionSelectName: undefined,
+  precisionSelectOnChange: undefined,
+  onsetDate: moment().format('YYYY-MM-DD'),
+  onsetDateInputName: undefined,
+  onsetDateInputOnChange: undefined,
+};
+
+const contextTypes = {
+  /* eslint-disable consistent-return */
+  intl: (context) => {
+    if (context.intl === undefined) {
+      return new Error('Please add locale prop to Base component to load translations');
+    }
+  },
+};
+
+class OnsetPicker extends React.Component {
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      granularity: this.props.granularity,
+      precision: this.props.precision,
+      onsetDate: this.props.onsetDate,
+      age: 1,
+      ageUnit: 'weeks',
+    };
+
+    // This binding is necessary to make `this` work in the callback
+    this.changeGranularity = this.changeGranularity.bind(this);
+    this.changePrecision = this.changePrecision.bind(this);
+    this.changeAge = this.changeAge.bind(this);
+    this.changeAgeUnit = this.changeAgeUnit.bind(this);
+    this.changeYear = this.changeYear.bind(this);
+    this.changeMonth = this.changeMonth.bind(this);
+    this.changeDate = this.changeDate.bind(this);
+    this.availableMonths = this.availableMonths.bind(this);
+    this.availableYears = this.availableYears.bind(this);
+    this.maxAgeCount = this.maxAgeCount.bind(this);
+    this.allowedAgeDurations = this.allowedAgeDurations.bind(this);
```

Comment:
I'm wondering if it would make sense to modify some of these functions to be static (availableMonths, availableYears, maxAgeCount, allowedAgeDurations) and move them into a separate Util file. It might make this code a little easier to read, over 400 lines of code in one file is a lot to look through.

Reply:
I've moved all the functions not related to onChange handling to a OnsetUtil file.